File: test/src/go/t07_socket.go
package main

import "syscall"

func main() {
	fds, _ := syscall.Socketpair(syscall.AF_UNIX, syscall.SOCK_STREAM, 0)
	buf := []byte{'A'}

	_ = syscall.Sendto(fds[0], buf, 0, nil)
	_, _, _ = syscall.Recvfrom(fds[1], buf, 0)

	_ = syscall.Shutdown(fds[0], syscall.SHUT_RDWR)
	_ = syscall.Close(fds[0])
	_ = syscall.Close(fds[1])
}

File: test/src/cpp/t07_socket.cpp
#include <unistd.h>
#include <sys/socket.h>

class Fd {
    int fd_;
public:
    explicit Fd(int fd = -1) : fd_(fd) {}
    ~Fd() { if (fd_ >= 0) (void)::close(fd_); }
    int get() const { return fd_; }
    void set(int fd) { fd_ = fd; }
    Fd(const Fd&) = delete;
    Fd& operator=(const Fd&) = delete;
};

class SocketPairTest {
public:
    static int run() {
        int fds_raw[2];
        char b[1];

        (void)::socketpair(AF_UNIX, SOCK_STREAM, 0, fds_raw);
        Fd a(fds_raw[0]);
        Fd bfd(fds_raw[1]);

        b[0] = 'A';
        (void)::sendto(a.get(), b, 1, 0, nullptr, 0);
        (void)::recvfrom(bfd.get(), b, 1, 0, nullptr, nullptr);
        (void)::shutdown(a.get(), SHUT_RDWR);

        return 0;
    }
};

int main() {
    return SocketPairTest::run();
}
